### DISCUSSION TITLE
Entity relations not updated during same kernel request

### DISCUSSION THREAD

**Discussion by kswzr** [doctrine/doctrine2#8622] [Support Questions]
Hi,

i have two entity objects: Product, ProductComponent. One product can have multiple components.
First creating the Product entity, persisting and flushing with the EntityManager.
Second, creating product component, but without knowledge about the real Product object, just having the id of the product:

```php
<?php
// run kernel..

// Product was persisted before in another class, during same kernel request

$em = ...

$productId = 'some-uuid';

$productReference = $em->getReference(Product::class, $productId);

$productComponent = new ProductComponent();
$productComponent->setProduct($productReference);

$em->persist($productComponent);
$em->flush();

$productRepository = $em->getRepository(Product::class);
$product = $productRepository->find($productId);

$components = $product->getProductComponents();
// $components will be empty ArrayCollection

// but what works:
$componentRepository = $em->getRepository(ProductComponent::class);
$components = $componentRepository->findBy(['product' => $productId]);

// $components will be found with the provided $productId

// end kernel..
```

When I am accessing now the `$productRepository` from a new kernel request, the returned product will have the component relations.

What's the reason for that? Is something missing?

**Comment by beberlei** [comment]
It took me a while to see this, `ProductComponent` and `Product` are bi-directional? I assume your code `$productComponent->setProduct()` does not update the bidirectional collection? The implementation must be:

```php
public function setProduct(Product $product)
{
    $this->product = $product;
    $product->addComponent($this);
}
// and Product:
public function addComponent(ProductComponent $component)
{
    $this->productComponents[] = $component;
}
```

If you are not updating this, then the call to `ProductRepository::find($productId)`` will return the reference that you loaded before and that was initialized.

Another way to fix this is to detach the reference after using:

```php
em->persist($productComponent);
$em->flush();
$em->detach($product);
```
